File: backend/scripts/check_critical_coverage.py
from __future__ import annotations

import sys
import xml.etree.ElementTree as ET
from pathlib import Path


THRESHOLDS: dict[str, float] = {
    "bootstrap.py": 80.0,
    "main.py": 70.0,
    "core/security.py": 70.0,
}


def main() -> int:
    coverage_file = Path("coverage.xml")
    if not coverage_file.exists():
        print(f"[coverage] Arquivo não encontrado: {coverage_file}", file=sys.stderr)
        return 1

    root = ET.parse(coverage_file).getroot()
    failures: list[str] = []

    for filename, minimum in THRESHOLDS.items():
        class_node = root.find(f".//class[@filename='{filename}']")
        if class_node is None:
            failures.append(f"[coverage] Arquivo crítico sem cobertura registrada: {filename}")
            continue

        line_rate = float(class_node.attrib.get("line-rate", "0"))
        percent = line_rate * 100
        if percent < minimum:
            failures.append(
                f"[coverage] {filename} abaixo do mínimo: {percent:.2f}% < {minimum:.0f}%"
            )
            continue

        print(f"[coverage] {filename}: {percent:.2f}% (mínimo {minimum:.0f}%)")

    if failures:
        for failure in failures:
            print(failure, file=sys.stderr)
        return 1

    print("[coverage] Cobertura crítica do backend validada com sucesso.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: frontend/scripts/check-critical-coverage.mjs
import fs from 'node:fs'
import path from 'node:path'

const coverageFile = path.resolve('coverage/lcov.info')

const thresholds = [
  { file: 'src/components/ConnectionStatus.tsx', min: 80 },
  { file: 'src/components/GlobalCommandPalette.tsx', min: 75 },
  { file: 'src/hooks/useSessionGuard.ts', min: 70 },
  { file: 'src/hooks/usePwaUpdatePrompt.ts', min: 70 },
  { file: 'src/lib/offlineQueue.ts', min: 65 },
]

function parseLcov(content) {
  const entries = []
  let current = null

  for (const rawLine of content.split('\n')) {
    const line = rawLine.trim()
    if (!line) continue

    if (line.startsWith('SF:')) {
      current = {
        file: line.slice(3),
        found: 0,
        hit: 0,
      }
      continue
    }

    if (line.startsWith('DA:') && current) {
      const [, hitCount] = line.slice(3).split(',')
      current.found += 1
      if (Number(hitCount) > 0) {
        current.hit += 1
      }
      continue
    }

    if (line === 'end_of_record' && current) {
      entries.push(current)
      current = null
    }
  }

  return entries
}

if (!fs.existsSync(coverageFile)) {
  console.error(`[coverage] Arquivo não encontrado: ${coverageFile}`)
  process.exit(1)
}

const content = fs.readFileSync(coverageFile, 'utf8')
const coverageEntries = parseLcov(content)

const failures = thresholds.flatMap((threshold) => {
  const entry = coverageEntries.find((candidate) => candidate.file.endsWith(threshold.file))

  if (!entry) {
    return [`[coverage] Arquivo crítico sem cobertura registrada: ${threshold.file}`]
  }

  const percent = entry.found === 0 ? 0 : (entry.hit / entry.found) * 100
  if (percent < threshold.min) {
    return [
      `[coverage] ${threshold.file} abaixo do mínimo: ${percent.toFixed(2)}% < ${threshold.min}%`,
    ]
  }

  console.log(`[coverage] ${threshold.file}: ${percent.toFixed(2)}% (mínimo ${threshold.min}%)`)
  return []
})

if (failures.length > 0) {
  for (const failure of failures) {
    console.error(failure)
  }
  process.exit(1)
}

console.log('[coverage] Cobertura crítica validada com sucesso.')
